apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+
+- Cohesion means all the code in a library is related to the other code in the
+  library, i.e. you don't combine code with completely orthogonal concerns into
+  the same library.
+- Coupling refers to implicit dependencies between seemingly unrelated peices of
+  code. It is the primary source of spaghetti, and is something we should strive
+  to minimize in our code base. A great example of coupling is having one module
+  interact directly with the DOM that another unrelated module created.
+- Encapsulation refers to the way in which data and functionality are (or are
+  not) exposed outside a given library. Libraries should expose their data and
+  functionality via a well-defined API that dictates how libraries can interact
+  with each other.
+
+In defining library bounderies, we should strive for a high degree of cohesion
+within a library, little if any implicit coupling with other libraries, and
+strong encapsulation.
+
+Because of the way our product is currently implemented, and because of how web
+technologies work at a fundamental level, there are limitations to how
+encapsulated, cohesive, and uncoupled we can make various libraries. For
+example, any code that renders directly to the DOM and relies on that DOM
+remaining in a particular state, is fundamentally not encapsulated because any
+other code on the page can do whatever it wants with that DOM at any
+time. Libraries like react and redux try to solve this by encouraging specific
+coding patterns that limit the surface area of unencapsulated code, but it is up
+to individual developers to actually stick to those patterns when writing code.
+
+For instances where react and redux do not provide sufficiently well defined
+patterns for our needs, we need to come up with our own patterns that reduce our
+exposure to unencapsulated spaghetti code.
+
+Getting back to "library types", I think there are two different library types
+for which we need to establish well defined patterns.
+
+1. **feature library** - This is a library that provide some meaty feature
+   of a page. It is characterized by the following attributes:
+
+   - maintains UI represented in the DOM (i.e. calls to ReactDOM.render(), or
+     use of jQuery to manipulate the DOM)
+   - interacts with a remote, stateful service (i.e. ajax calls that generate
+     writes to a database)
+   - uses or provides shared application state
+
+   Some examples of "feature libraries" in our code base today would be the code
+   found in `src/applab`, `src/netsim`, and `src/flappy`. Note however that
+   these directories also contain `main.js` files which strictly speaking are
+   not part of the "library" but are actually entry points that get loaded on a
+   particular html page.
+
+2. **capsule library** - This is a library that is wholy encapsulated and
+   decoupled, and which therefore:
+
+   - does not maintain any UI represented in the DOM (react components are OK, but
+     calls to ReactDOM.render() are not)
```

Comment:
Maybe also describe the non-React analogue to this?  Many of our libraries (like Blockly) do completely take over part of the DOM, but don't need to interact with an API or external application state.  Does that satisfy this requirement?  Is DOM manipulation okay as long as the parent somehow injects part of the DOM into the library and the library doesn't do anything outside of that subtree?